appsrc/dcfile/rawftodc.man
.TH RAWFTODC 1 "15 October 2002" "DICOM PS3" "Create DICOM from raw float file"
.SH NAME
rawftodc \- ACR/NEMA DICOM PS3 ... Create DICOM from raw float file
.SH SYNOPSIS
.HP 10
.B rawftodc
.B \-rows|\-height|\-h " rows"
.B \-columns|\-width|\-w " cols"
[
.B \-frames " frames"
]
[
.B \-scale " scalefactor"
]
[
.B \-componentoffset " 0|1"
]
[
.B \-componentnumber " 2"
]
[
.B \-little|\-l|\-big
]
[
.B \-samples " samples"
]
[
.B \-gray|\-grey|\-rgb|\-argb|\-hsv|\-cmyk
]
[
.B \-photometric-interpretation " xxx"
]
[
.B \-color-by-pixel|\-color-by-plane
]
[
.B \-v|verbose
]
.so man1/genin.so
.so man1/optout.so
.SH DESCRIPTION
.LP
.B rawftodc
reads the named raw input file and converts it to a 16-bit DICOM Secondary Capture Image Storage SOP Class instance stored in a DICOM file. Only uncompressed raw images with one sample per byte or word are supported, though multiple samples per pixel are allowed. The planar configuration can be specified but not re-ordered.
.LP
During the translation, specific attributes may be deleted, added or
replaced, class and instance unique identifiers generated or removed,
group lengths added, or private attributes removed, as specified by
the replacement options described in dcintro(1). This allows the
specification of additional DICOM attributes, such as the patient's
name, necessary to make a complete and useful SC instance.
.SH OPTIONS
The verbose output goes to standard error.
.PP
.PP
The basic switches are described in dcintro(1). Options specific to this program are:
.TP
.BI \-rows|\-height|\-h " rows"
.RS
Required.
.RE
.TP
.BI \-columns|\-width|\-w " cols"
.RS
Required.
.RE
.TP
.BI \-little|\-big
.RS
Required.
.RE
.TP
.BI \-frames " frames"
.RS
Defaults to 1.
.RE
.TP
.BI \-scale " scalefactor"
.RS
Normally the floating point values are truncated to 16 bit integers to create pixel 
values. A scale factor may be specified by which to multiply them before truncation.
Defaults to 1.
.RE
.TP
.BI \-componentoffset " 0|1"
.RS
Normally it is assumed there is one floating point value per pixel. When more than one
component is specified by -componentnumber, then which component (first, second etc.)
may be specified. Usually the first component of a complex pair is real, and the
second imaginary. Defaults to 0 (first component).
.RE
.TP
.BI \-componentnumber " 2"
.RS
Specifies that there is more than one floating point value per pixel, e.g. a complex pair
of two components. Defaults to 1 (one component).
.RE
.TP
.BI \-samples " samples"
.RS
Defaults to appropriate value for photometric interpretation, or 1 if none specified. Only needs to be specified for unusual photometric interpretations.
.RE
.TP
.BI \-gray|\-grey|\-rgb|\-argb|\-hsv|\-cmyk
.RS
Recognized photometric interpretations, defaults to gray if no samples specified or samples is 1, and rgb if samples is 3.
.RE
.TP
.BI \-photometric-interpretation " xxx"
.RS
Explicitly specify string value for photometric interpretation attribute; will need value for samples.
.RE
.TP
.BI \-color-by-pixel|\-color-by-plane
.RS
Specifies the planar configuration. Required if samples not 1 (explicitly or deduced from photometric interpretation). No default.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dccp(1) ,
.BR dctoraw(1) ,
.BR pnmtodc(1) ,
.BR rawtodc(1) ,
.BR dciodvfy(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/rawftodc.cc
#include "attrtype.h"
#include "attrothr.h"
#include "attrmxls.h"
#include "transynu.h"
#include "elmconst.h"
#include "mesgtext.h"
#include "ioopt.h"
#include "dcopt.h"
#include "srcsink.h"
#include "convtype.h"
#include "fltype.h"

class RawFloat_PixelDataSourceBase : public SourceBase<Uint16> {
protected:
	istream *istr;
	long offset;
	Uint16 rows;
	Uint16 columns;
	Uint16 frames;
	Uint16 samples;
	double scaleFactor;
	Uint16 componentOffset;
	Uint16 componentNumber;

	Uint32 totalrows;
	Uint32 row;
	Uint16 col;
	Uint16 frame;
	Uint16 *buffer;
public:
	RawFloat_PixelDataSourceBase(istream& i,long off,Uint16 r,Uint16 c,
		Uint16 f,Uint16 s,double scf,Uint16 cOffset,Uint16 cNumber)
			: SourceBase<Uint16>()
		{
			istr=&i;
			offset=off;
			rows=r;
			Assert(rows);
			columns=c;
			Assert(columns);
			frames=f;
			Assert(frames);
			samples=s;
			Assert(samples);
			scaleFactor=scf;
			componentOffset=cOffset;
			componentNumber=cNumber;

			totalrows=((Uint32)rows)*frames;
			row=0;
			col=0;
			frame=0;
			buffer=new Uint16[c*s];
			Assert(buffer);
		}

	virtual ~RawFloat_PixelDataSourceBase()
		{
			if (buffer) delete[] buffer;
		}

	virtual size_t read(void) = 0;

	const Uint16 *getBuffer(void)		{ return buffer; }

	size_t getBufferCount(void) const	{ return col*samples; }

	int good(void) const	{ return istr && istr->good() && row < totalrows; }
};

class RawFloat_PixelDataSource_LittleEndian : public RawFloat_PixelDataSourceBase {
public:
	RawFloat_PixelDataSource_LittleEndian(istream& i,long off,Uint16 r,Uint16 c,
		Uint16 f,Uint16 s,double scf,Uint16 cOffset,Uint16 cNumber)
			: RawFloat_PixelDataSourceBase(i,off,r,c,f,s,scf,cOffset,cNumber)
		{}

	size_t read(void)
		{
			if (row == 0 && frame == 0) {
				Assert(istr);
				istr->seekg(offset,ios::beg);
			}
			col=0;
			if (!istr->good() || row++ >= totalrows) return 0;
			Uint16 *ptr=buffer;
			while (col < columns) {
				Uint16 sample=0;
				while (sample < samples) {
					int component;
					IEEE_Float32_L value;
					unsigned char *bytes = (unsigned char *)(&value);
					for (component=0; component<componentOffset; ++component) {		// skip components before the one we want
						istr->read((char *)bytes,4);
						if (!istr->good()) return 0;
					}
					istr->read((char *)bytes,4);
					if (!istr->good()) return 0;
					Uint16 pixel = (Uint16)(value * scaleFactor);
//cerr << value << ",";
					*ptr++=pixel; ++sample;
					for (component=componentOffset+1; component<componentNumber; ++component) {	// skip components after the one we want
						istr->read((char *)bytes,4);
						if (!istr->good()) return 0;
					}
				}
				++col;
			}
//cerr << endl;
			return col*samples;
		}
};

class RawFloat_PixelDataSource_BigEndian : public RawFloat_PixelDataSourceBase {
public:
	RawFloat_PixelDataSource_BigEndian(istream& i,long off,Uint16 r,Uint16 c,
		Uint16 f,Uint16 s,double scf,Uint16 cOffset,Uint16 cNumber)
			: RawFloat_PixelDataSourceBase(i,off,r,c,f,s,scf,cOffset,cNumber)
		{}

	size_t read(void)
		{
			if (row == 0 && frame == 0) {
				Assert(istr);
				istr->seekg(offset,ios::beg);
			}
			col=0;
			if (!istr->good() || row++ >= totalrows) return 0;
			Uint16 *ptr=buffer;
			while (col < columns) {
				Uint16 sample=0;
				while (sample < samples) {
					int component;
					IEEE_Float32_B value;
					unsigned char *bytes = (unsigned char *)(&value);
					for (component=0; component<componentOffset; ++component) {		// skip components before the one we want
						istr->read((char *)bytes,4);
						if (!istr->good()) return 0;
					}
					istr->read((char *)bytes,4);
					if (!istr->good()) return 0;
					Uint16 pixel = (Uint16)(value * scaleFactor);
//cerr << value << ",";
					*ptr++=pixel; ++sample;
					for (component=componentOffset+1; component<componentNumber; ++component) {	// skip components after the one we want
						istr->read((char *)bytes,4);
						if (!istr->good()) return 0;
					}
				}
				++col;
			}
//cerr << endl;
			return col*samples;
		}
};

int
main(int argc,char **argv)
{
	bool bad=false;

	GetNamedOptions 	options(argc,argv);
	InputOptions		input_options(options);
	DicomOutputOptions 	dicom_output_options(options);

	bool verbose=options.get("verbose") || options.get("v");

	long offset=0;
	(void)options.get("offset",offset);

	unsigned rows=0;
	if (!options.get("rows",rows) && !options.get("height",rows) && !options.get("h",rows)) {
		cerr << EMsgDC(NeedOption) << " - rows" << endl;
		bad=true;
	}
	
	unsigned cols=0;
	if (!options.get("columns",cols) && !options.get("width",cols) && !options.get("w",cols)) {
		cerr << EMsgDC(NeedOption) << " - columns" << endl;
		bad=true;
	}
	
	unsigned frames=1;
	options.get("frames",frames);
	
	// e.g. for real of real and imaginary pair, componentOffset=0,componentNumber=2
	
	unsigned componentOffset=0;
	options.get("componentoffset",componentOffset);
	
	unsigned componentNumber=1;
	options.get("componentnumber",componentNumber);
	
	double scaleFactor=1;
	options.get("scale",scaleFactor);	// use 65535 to convert 0.0-1.0 to 16 but unsigned
	
	unsigned bits=16;

	bool littleendian=options.get("little") || options.get("l");
	bool bigendian=options.get("big") || options.get("b");

	if (littleendian && bigendian) {
		cerr << EMsgDC(OptionsIncompatible) << " - big & little" << endl;
		bad=true;
	}

	unsigned samples;
	bool samplespresent=options.get("samples",samples);

	const char *photometricInterpretation=0;

	int graypresent=(options.get("gray") || options.get("grey")) ? 1 : 0;
	int rgbpresent=options.get("rgb") ? 1 : 0;
	int argbpresent=options.get("argb") ? 1 : 0;
	int hsvpresent=options.get("hsv") ? 1 : 0;
	int cmykpresent=options.get("cmyk") ? 1 : 0;
	int photometricpresent=options.get("photometric-interpretation",photometricInterpretation) ? 1 : 0;

	int sumofphotometric=graypresent+rgbpresent+argbpresent+hsvpresent+cmykpresent+photometricpresent;

	if (sumofphotometric == 0) {
		if (samplespresent) {
			if (samples == 1) {
				photometricInterpretation="MONOCHROME2";
			}
			else if (samples == 3) {
				photometricInterpretation="RGB";
			}
		}
		else {
			photometricInterpretation="MONOCHROME2";
			samples=1;
		}
	}
	else if (sumofphotometric > 1) {
		cerr << EMsgDC(OptionsIncompatible) << " - gray,rgb,argb,hsv,cmyk,photometric-interpretation" << endl;
		bad=true;
	}
	else {
		if (graypresent) {
			if (samplespresent && samples != 1) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 1 & gray" << endl;
				bad=true;
			}
			else
				samples=1;
			photometricInterpretation="MONOCHROME2";
		}
		else if (rgbpresent) {
			if (samplespresent && samples != 3) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 3 & rgb" << endl;
				bad=true;
			}
			else
				samples=3;
			photometricInterpretation="RGB";
		}
		else if (argbpresent) {
			if (samplespresent && samples != 4) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 4 & argb" << endl;
				bad=true;
			}
			else
				samples=4;
			photometricInterpretation="ARGB";
		}
		else if (hsvpresent) {
			if (samplespresent && samples != 3) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 3 & hsv" << endl;
				bad=true;
			}
			else
				samples=3;
			photometricInterpretation="HSV";
		}
		else if (cmykpresent) {
			if (samplespresent && samples != 4) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 4 & cmyk" << endl;
				bad=true;
			}
			else
				samples=4;
			photometricInterpretation="CMYK";
		}
		else if (photometricpresent) {
			if (!samplespresent) {
				cerr << EMsgDC(OptionRequired) << " - samples" << endl;
				bad=true;
			}
		}
	}

	Assert(photometricInterpretation);
	Assert(samples);

	bool bypixelpresent=options.get("color-by-pixel");
	bool byplanepresent=options.get("color-by-plane");

	unsigned planarConfiguration;

	if (bypixelpresent && byplanepresent) {
		cerr << EMsgDC(OptionsIncompatible) << " - color-by-pixel & color-by-plane" << endl;
		bad=true;
	}
	if (samples == 1) {
		if (bypixelpresent || byplanepresent) {
			cerr << EMsgDC(OptionsIncompatible) << " - (color-by-pixel|color-by-plane) & samples = 1" << endl;
			bad=true;
		}
	}
	else {
		if (bypixelpresent)
			planarConfiguration=0;
		else if (byplanepresent)
			planarConfiguration=1;
		else {
			cerr << EMsgDC(OptionRequired) << " - color-by-pixel | color-by-plane" << endl;
			bad=true;
		}
	}
			
	input_options.done();
	dicom_output_options.done();
	options.done();

	InputOpenerFromOptions input_opener(
		options,input_options.filename,cin);
	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << input_options.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!input_options.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options
	 || bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< input_options.usage()
			<< dicom_output_options.usage()
			<< " -rows|-height|-h rows"
			<< " -columns|-width|-w columns"
			<< " [-frames frames]"
			<< " [-scale scalefactor]"
			<< " [-componentoffset 0|1]"
			<< " [-componentnumber 2]"
			<< " [-little|-l|-big]"
			<< " [-samples n]"
			<< " [-offset bytes]"
			<< " [[-gray|-grey|-rgb|-argb|-hsv|-cmyk]|[-photometric-interpretation xxx]]"
			<< " [-color-by-pixel|-color-by-plane]"
			<< " [-v|-verbose]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	istream in(input_opener);
	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	TransferSyntax transfersyntax(
		dicom_output_options.transfersyntaxuid
		? dicom_output_options.transfersyntaxuid
		: DefaultTransferSyntaxUID);

	TextOutputStream log (cerr);
	if (verbose) log << "Converting ..." << endl;

	Uint16 bitsAllocated=((bits-1)/8+1)*8;
	Uint16 bitsStored=bits;
	Uint16 highBit=bits-1;
	//Uint16 frames=1;

	list+=new UnsignedShortAttribute(TagFromName(BitsAllocated),bitsAllocated);
	list+=new UnsignedShortAttribute(TagFromName(BitsStored),bitsStored);
	list+=new UnsignedShortAttribute(TagFromName(HighBit),highBit);
	list+=new UnsignedShortAttribute(TagFromName(Rows),rows);
	list+=new UnsignedShortAttribute(TagFromName(Columns),cols);
	list+=new IntegerStringAttribute(TagFromName(NumberOfFrames),(Uint16)frames);
	list+=new UnsignedShortAttribute(TagFromName(SamplesPerPixel),samples);
	list+=new UnsignedShortAttribute(TagFromName(PixelRepresentation),0u);
	list+=new CodeStringAttribute(TagFromName(PhotometricInterpretation),photometricInterpretation);
	if (samples > 1) list+=new UnsignedShortAttribute(TagFromName(PlanarConfiguration),planarConfiguration);

	// various Type 1 and Type 2 attributes for mandatory SC modules ...

	list+=new PersonNameAttribute(TagFromName(PatientName),"^^^^");
	list+=new LongStringAttribute(TagFromName(PatientID));
	list+=new DateStringAttribute(TagFromName(PatientBirthDate));
	list+=new CodeStringAttribute(TagFromName(PatientSex));
	list+=new ShortStringAttribute(TagFromName(StudyID));
	list+=new DateStringAttribute(TagFromName(StudyDate));
	list+=new TimeStringAttribute(TagFromName(StudyTime));
	list+=new PersonNameAttribute(TagFromName(ReferringPhysicianName),"^^^^");
	list+=new ShortStringAttribute(TagFromName(AccessionNumber));
	list+=new IntegerStringAttribute(TagFromName(SeriesNumber));
	list+=new IntegerStringAttribute(TagFromName(InstanceNumber));

	list+=new CodeStringAttribute(TagFromName(Modality),"OT");
	list+=new CodeStringAttribute(TagFromName(ConversionType),"WSD");

	list+=new LongStringAttribute(TagFromName(Manufacturer));
	list+=new CodeStringAttribute(TagFromName(PatientOrientation));

	RawFloat_PixelDataSourceBase *pixeldatasrc=0;
	if (littleendian) {
		pixeldatasrc=new RawFloat_PixelDataSource_LittleEndian(in,offset,rows,cols,frames,samples,scaleFactor,componentOffset,componentNumber);
	}
	else if (bigendian) {
		pixeldatasrc=new RawFloat_PixelDataSource_BigEndian(in,offset,rows,cols,frames,samples,scaleFactor,componentOffset,componentNumber);
	}
	Assert(pixeldatasrc);

	list+=new OtherUnspecifiedLargeAttribute(
		TagFromName(PixelData),
		pixeldatasrc,
		rows,
		cols,
		frames,
		samples,
		&transfersyntax,
		0,bitsAllocated,bitsStored,highBit);

	if (!usualManagedAttributeListWrite(list,dout,
		dicom_output_options,log,verbose)) return 1;

	return 0;
}

